### DISCUSSION TITLE
Delete site

### DISCUSSION THREAD

**Discussion by andershagman** [netbox-community/netbox#8299] [Q&A]
Do anyone have script to delete a site with its content or other idea?
Most of the objects can be deleted rather easily through the gui but not ip addresses as you can´t list them with site filter.
But all in all it´s a lot of steps to delete a site. I listed 13 steps.

**Comment by candlerb** [comment]
> Most of the objects can be deleted rather easily through the gui but not ip addresses as you can´t list them with site filter.

That's because there is no direct relationship between IP address and site - see the [model](https://github.com/netbox-community/netbox/blob/v3.1.5/netbox/ipam/models/ip.py#L743-L820).

One way that an IP address maybe associated _indirectly_ with a site is through an interface to a device, or a vminterface to a virtual machine.  A device has a mandatory, direct association with a site, so that's straightforward. However, a virtualmachine is only indirectly related to a site; it must be linked to a Cluster, but the Cluster-Site association is optional.  The cluster may contain multiple devices, but they could be in different sites.

You should find that if you delete a device or VM, this will automatically delete its interfaces and any IP addresses attached to those interfaces.  (At least I believe this works through the GUI; I have not tested via the REST API)

However, not all IP addresses are linked to an interface.

There are other ways that an IP address can be indirectly related to a site: for example, the enclosing prefix _may_ be linked to a site; and/or the prefix _may_ be linked to a VLAN, and that VLAN _may_ be associated to a site.  Deleting a prefix does *not* automatically delete all the IP addresses within it (and you'd have to be careful if you did, since prefixes can also enclose other prefixes).  You can, however, easily find all the IP addresses within a prefix, and delete them.

> But all in all it's a lot of steps to delete a site. I listed 13 steps.

Why don't you list them here, so others can comment?  Your 13 steps may not be appropriate for all users, but you can implement those 13 steps in a [Custom Script](https://netbox.readthedocs.io/en/stable/customization/custom-scripts/) if you wish.  There are some example custom scripts you can base your work on [here](https://github.com/netbox-community/reports/tree/master/scripts), and you could share your script when done in the same repository.

**Reply by andershagman** [reply]
Thanks for a quick answer.

Yes the lack of an easy way to filter ip addresses on site or on miltiple parent prefixes makes the necessary steps to delete adresses massive.
The big clicking part is repeating step 2-4. 

13 steps of gui clicking:
1.	Go to site
2.	choose prefixes in stats view
3.	Click on one prefix and select IP adresses tab
4.	Mark all addresses and choose delete under the list
5.	repeate step 2-4 until all addresses are deleted
6.	Mark all prefix belonging to the site and choose delete
7.	Go to site and choose vlan in stats view
8.	Mark the vlan and click delete.
9.	Go to site and choose devices in stats view
10.	Mark all devices and click delete
11.	Go to site and choose circuits in stats view
12.	Mark and delete
13.	Go to, an hopfully emty, site and delete it

Here is a change logg of the manual gui removal of one small typical site (remote office).

|                Time | Action  | Type                        | Object                              |
| ------------------- | ------- | --------------------------- | ----------------------------------- |
| 2022-01-10 10:14:00 | Deleted | ipam.ipaddress              | x.y.222.1/24                        |
| 2022-01-10 10:14:00 | Deleted | ipam.ipaddress              | x.y.1.122/30                        |
| 2022-01-10 10:14:00 | Deleted | ipam.ipaddress              | x.y.1.121/30                        |
| 2022-01-10 10:15:00 | Deleted | ipam.ipaddress              | x.y.9.122/30                        |
| 2022-01-10 10:15:00 | Deleted | ipam.ipaddress              | x.y.9.121/30                        |
| 2022-01-10 10:16:00 | Deleted | ipam.ipaddress              | x.y.222.1/24                        |
| 2022-01-10 10:17:00 | Deleted | ipam.ipaddress              | x.y.147.193/27                      |
| 2022-01-10 10:17:00 | Deleted | ipam.ipaddress              | x.y.9.122/30                        |
| 2022-01-10 10:17:00 | Deleted | ipam.ipaddress              | x.y.9.121/30                        |
| 2022-01-10 10:17:00 | Deleted | ipam.ipaddress              | x.y.1.122/30                        |
| 2022-01-10 10:17:00 | Deleted | ipam.ipaddress              | x.y.1.121/30                        |
| 2022-01-10 10:18:00 | Deleted | ipam.ipaddress              | x.y.222.1/24                        |
| 2022-01-10 10:18:00 | Deleted | ipam.ipaddress              | x.y.9.122/30                        |
| 2022-01-10 10:18:00 | Deleted | ipam.ipaddress              | x.y.9.121/30                        |
| 2022-01-10 10:18:00 | Deleted | ipam.ipaddress              | x.y.1.122/30                        |
| 2022-01-10 10:18:00 | Deleted | ipam.ipaddress              | x.y.1.121/30                        |
| 2022-01-10 10:19:00 | Deleted | ipam.ipaddress              | x.y.1.122/30                        |
| 2022-01-10 10:19:00 | Deleted | ipam.ipaddress              | x.y.1.121/30                        |
| 2022-01-10 10:20:00 | Deleted | ipam.ipaddress              | x.y.9.122/30                        |
| 2022-01-10 10:20:00 | Deleted | ipam.ipaddress              | x.y.9.121/30                        |
| 2022-01-10 10:20:00 | Deleted | ipam.ipaddress              | x.y.1.122/30                        |
| 2022-01-10 10:20:00 | Deleted | ipam.ipaddress              | x.y.1.121/30                        |
| 2022-01-10 10:20:00 | Deleted | ipam.ipaddress              | x.y.41.200/24                       |
| 2022-01-10 10:20:00 | Deleted | ipam.ipaddress              | x.y.41.180/24                       |
| 2022-01-10 10:20:00 | Deleted | ipam.ipaddress              | x.y.41.120/24                       |
| 2022-01-10 10:20:00 | Deleted | ipam.ipaddress              | x.y.41.1/24                         |
| 2022-01-10 10:20:00 | Deleted | ipam.ipaddress              | x.y.9.122/30                        |
| 2022-01-10 10:20:00 | Deleted | ipam.ipaddress              | x.y.9.121/30                        |
| 2022-01-10 10:21:00 | Deleted | ipam.ipaddress              | x.y.9.122/30                        |
| 2022-01-10 10:21:00 | Deleted | ipam.ipaddress              | x.y.9.121/30                        |
| 2022-01-10 10:21:00 | Deleted | ipam.ipaddress              | x.y.1.122/30                        |
| 2022-01-10 10:21:00 | Deleted | ipam.ipaddress              | x.y.1.121/30                        |
| 2022-01-10 10:21:00 | Deleted | ipam.ipaddress              | x.y.222.1/24                        |
| 2022-01-10 10:22:00 | Deleted | ipam.ipaddress              | x.y.222.1/24                        |
| 2022-01-10 10:22:00 | Deleted | ipam.ipaddress              | x.y.11.193/27                       |
| 2022-01-10 10:22:00 | Deleted | ipam.ipaddress              | x.y.9.122/30                        |
| 2022-01-10 10:22:00 | Deleted | ipam.ipaddress              | x.y.9.121/30                        |
| 2022-01-10 10:22:00 | Deleted | ipam.ipaddress              | x.y.1.122/30                        |
| 2022-01-10 10:22:00 | Deleted | ipam.ipaddress              | x.y.1.121/30                        |
| 2022-01-10 10:23:00 | Deleted | ipam.ipaddress              | x.y.9.122/30                        |
| 2022-01-10 10:23:00 | Deleted | ipam.ipaddress              | x.y.9.121/30                        |
| 2022-01-10 10:23:00 | Deleted | ipam.ipaddress              | x.y.1.122/30                        |
| 2022-01-10 10:23:00 | Deleted | ipam.ipaddress              | x.y.1.121/30                        |
| 2022-01-10 10:24:00 | Deleted | ipam.prefix                 | x.y.9.120/30                        |
| 2022-01-10 10:24:00 | Deleted | ipam.prefix                 | x.y.1.120/30                        |
| 2022-01-10 10:24:00 | Deleted | ipam.prefix                 | x.y.222.0/24                        |
| 2022-01-10 10:24:00 | Deleted | ipam.prefix                 | x.y.11.192/27                       |
| 2022-01-10 10:24:00 | Deleted | ipam.prefix                 | x.y.9.120/30                        |
| 2022-01-10 10:24:00 | Deleted | ipam.prefix                 | x.y.1.120/30                        |
| 2022-01-10 10:24:00 | Deleted | ipam.prefix                 | x.y.9.120/30                        |
| 2022-01-10 10:24:00 | Deleted | ipam.prefix                 | x.y.1.120/30                        |
| 2022-01-10 10:24:00 | Deleted | ipam.prefix                 | x.y.222.0/24                        |
| 2022-01-10 10:24:00 | Deleted | ipam.prefix                 | x.y.9.120/30                        |
| 2022-01-10 10:24:00 | Deleted | ipam.prefix                 | x.y.1.120/30                        |
| 2022-01-10 10:24:00 | Deleted | ipam.prefix                 | x.y.41.0/24                         |
| 2022-01-10 10:24:00 | Deleted | ipam.prefix                 | x.y.9.120/30                        |
| 2022-01-10 10:24:00 | Deleted | ipam.prefix                 | x.y.1.120/30                        |
| 2022-01-10 10:24:00 | Deleted | ipam.prefix                 | x.y.222.0/24                        |
| 2022-01-10 10:24:00 | Deleted | ipam.prefix                 | x.y.9.120/30                        |
| 2022-01-10 10:24:00 | Deleted | ipam.prefix                 | x.y.1.120/30                        |
| 2022-01-10 10:24:00 | Deleted | ipam.prefix                 | x.y.147.192/27                      |
| 2022-01-10 10:24:00 | Deleted | ipam.prefix                 | x.y.9.120/30                        |
| 2022-01-10 10:24:00 | Deleted | ipam.prefix                 | x.y.1.120/30                        |
| 2022-01-10 10:24:00 | Deleted | ipam.prefix                 | x.y.222.0/24                        |
| 2022-01-10 10:24:00 | Deleted | ipam.prefix                 | x.y.222.0/24                        |
| 2022-01-10 10:24:00 | Deleted | ipam.prefix                 | x.y.9.120/30                        |
| 2022-01-10 10:24:00 | Deleted | ipam.prefix                 | x.y.1.120/30                        |
| 2022-01-10 10:25:00 | Deleted | dcim.device                 | lvk-0018-klt-l3-a                   |
| 2022-01-10 10:25:00 | Deleted | dcim.interface              | Vlan610                             |
| 2022-01-10 10:25:00 | Deleted | dcim.interface              | GigabitEthernet8.3926               |
| 2022-01-10 10:25:00 | Deleted | dcim.interface              | GigabitEthernet8.1926               |
| 2022-01-10 10:25:00 | Deleted | dcim.interface              | GigabitEthernet8.3927               |
| 2022-01-10 10:25:00 | Deleted | dcim.interface              | Vlan605                             |
| 2022-01-10 10:25:00 | Deleted | dcim.interface              | GigabitEthernet8.1935               |
| 2022-01-10 10:25:00 | Deleted | dcim.interface              | GigabitEthernet8.3935               |
| 2022-01-10 10:25:00 | Deleted | dcim.interface              | Vlan608                             |
| 2022-01-10 10:25:00 | Deleted | dcim.interface              | GigabitEthernet8.1927               |
| 2022-01-10 10:25:00 | Deleted | dcim.interface              | Vlan699                             |
| 2022-01-10 10:25:00 | Deleted | dcim.interface              | Vlan615                             |
| 2022-01-10 10:25:00 | Deleted | dcim.interface              | GigabitEthernet8.1923               |
| 2022-01-10 10:25:00 | Deleted | dcim.interface              | GigabitEthernet8.3923               |
| 2022-01-10 10:25:00 | Deleted | dcim.interface              | GigabitEthernet8.1921               |
| 2022-01-10 10:25:00 | Deleted | dcim.interface              | GigabitEthernet8.3921               |
| 2022-01-10 10:25:00 | Deleted | dcim.interface              | GigabitEthernet8.1937               |
| 2022-01-10 10:25:00 | Deleted | dcim.interface              | GigabitEthernet8.3937               |
| 2022-01-10 10:25:00 | Deleted | dcim.interface              | GigabitEthernet8.1924               |
| 2022-01-10 10:25:00 | Deleted | dcim.interface              | GigabitEthernet8.1925               |
| 2022-01-10 10:25:00 | Deleted | dcim.interface              | GigabitEthernet8.3925               |
| 2022-01-10 10:25:00 | Deleted | dcim.interface              | GigabitEthernet8.3924               |
| 2022-01-10 10:25:00 | Deleted | dcim.interface              | Loopback1                           |
| 2022-01-10 10:25:00 | Deleted | dcim.interface              | Vlan604                             |
| 2022-01-10 10:25:00 | Deleted | dcim.interface              | GigabitEthernet8                    |
| 2022-01-10 10:25:00 | Deleted | dcim.interface              | GigabitEthernet7                    |
| 2022-01-10 10:25:00 | Deleted | dcim.interface              | GigabitEthernet6                    |
| 2022-01-10 10:25:00 | Deleted | dcim.interface              | GigabitEthernet5                    |
| 2022-01-10 10:25:00 | Deleted | dcim.interface              | GigabitEthernet4                    |
| 2022-01-10 10:25:00 | Deleted | dcim.interface              | GigabitEthernet3                    |
| 2022-01-10 10:25:00 | Deleted | dcim.interface              | GigabitEthernet2                    |
| 2022-01-10 10:25:00 | Deleted | dcim.interface              | GigabitEthernet1                    |
| 2022-01-10 10:25:00 | Deleted | dcim.interface              | GigabitEthernet0                    |
| 2022-01-10 10:25:00 | Deleted | dcim.interface              | FastEthernet0                       |
| 2022-01-10 10:25:00 | Deleted | dcim.powerport              | PSU0                                |
| 2022-01-10 10:25:00 | Deleted | dcim.consoleport            | con0                                |
| 2022-01-10 10:25:00 | Deleted | dcim.device                 | lvk-0018-klt-cpe-a                  |
| 2022-01-10 10:25:00 | Deleted | dcim.rearport               | SFP                                 |
| 2022-01-10 10:25:00 | Deleted | dcim.cable                  | #446                                |
| 2022-01-10 10:25:00 | Deleted | dcim.cable                  | #445                                |
| 2022-01-10 10:25:00 | Deleted | dcim.frontport              | Po4                                 |
| 2022-01-10 10:25:00 | Deleted | dcim.frontport              | Po3                                 |
| 2022-01-10 10:25:00 | Deleted | dcim.frontport              | Po2                                 |
| 2022-01-10 10:25:00 | Deleted | dcim.frontport              | Po1                                 |
| 2022-01-10 10:25:00 | Deleted | ipam.vlan                   | link mt-d b-way sd19-c (3937)       |
| 2022-01-10 10:25:00 | Deleted | ipam.vlan                   | link lkl-a b-way sd19a (3935)       |
| 2022-01-10 10:25:00 | Deleted | ipam.vlan                   | link lkl-b b-way sd19-c (3927)      |
| 2022-01-10 10:25:00 | Deleted | ipam.vlan                   | link fastighet-a b-way sd19a (3926) |
| 2022-01-10 10:25:00 | Deleted | ipam.vlan                   | link fastighet-a b-way sd19a (3925) |
| 2022-01-10 10:25:00 | Deleted | ipam.vlan                   | link guest-a b-way sd19a (3924)     |
| 2022-01-10 10:25:00 | Deleted | ipam.vlan                   | link ip-tele-a b-way sd19a (3923)   |
| 2022-01-10 10:25:00 | Deleted | ipam.vlan                   | link mgmt-a b-way sd19a (3921)      |
| 2022-01-10 10:25:00 | Deleted | ipam.vlan                   | link mt-d a-way sd18-c (1937)       |
| 2022-01-10 10:25:00 | Deleted | ipam.vlan                   | link lkl-a a-way sd18a (1935)       |
| 2022-01-10 10:25:00 | Deleted | ipam.vlan                   | link lkl-b a-way sd18-c (1927)      |
| 2022-01-10 10:25:00 | Deleted | ipam.vlan                   | link fastighet-a a-way sd18a (1926) |
| 2022-01-10 10:25:00 | Deleted | ipam.vlan                   | link fastighet-a a-way sd18a (1925) |
| 2022-01-10 10:25:00 | Deleted | ipam.vlan                   | link guest-a a-way sd18a (1924)     |
| 2022-01-10 10:25:00 | Deleted | ipam.vlan                   | link ip-tele-a a-way sd18a (1923)   |
| 2022-01-10 10:25:00 | Deleted | ipam.vlan                   | link mgmt-a a-way sd18a (1921)      |
| 2022-01-10 10:25:00 | Deleted | ipam.vlan                   | mgmt-a (699)                        |
| 2022-01-10 10:25:00 | Deleted | ipam.vlan                   | mt-d (617)                          |
| 2022-01-10 10:25:00 | Deleted | ipam.vlan                   | klt-a (615)                         |
| 2022-01-10 10:25:00 | Deleted | ipam.vlan                   | ip-tele-a (610)                     |
| 2022-01-10 10:25:00 | Deleted | ipam.vlan                   | guest-a (608)                       |
| 2022-01-10 10:25:00 | Deleted | ipam.vlan                   | lkl-b (607)                         |
| 2022-01-10 10:25:00 | Deleted | ipam.vlan                   | fastighet-b (605)                   |
| 2022-01-10 10:25:00 | Deleted | ipam.vlan                   | fastighet-a (604)                   |
| 2022-01-10 10:26:00 | Deleted | dcim.site                   | LVK                                 |
| 2022-01-10 10:26:00 | Deleted | dcim.location               | Rum 01-0018                         |
| 2022-01-10 10:26:00 | Deleted | dcim.location               | Hus 01                              |
| 2022-01-10 10:26:00 | Deleted | circuits.circuit            | FNT-92603                           |
| 2022-01-10 10:26:00 | Deleted | circuits.circuittermination | Termination Z: Telia WAN            |
| 2022-01-10 10:26:00 | Deleted | circuits.circuittermination | Termination A: LVK                  |

**Reply by candlerb** [reply]
OK, so you're using the `IP >-*- Prefix >-- Site` association.

(*) the relation between IP and Prefix is not a direct foreign key link in the database (see #7845) which means it can't be used in a direct database join in the Django ORM - although maybe it could with CTE (see #6587).  In short, the filter you want isn't easy to implement in the current framework.

For comparison, you *can* filter Virtual Machines by Site - e.g. `/virtualization/virtual-machines/?site_id=1`.  Looking in the [source](https://github.com/netbox-community/netbox/blob/v3.1.5/netbox/virtualization/filtersets.py#L181-L185), it's using the relationship `VirtualMachine >-- Cluster >-- Site`:

```
    site_id = django_filters.ModelMultipleChoiceFilter(
        field_name='cluster__site',
        queryset=Site.objects.all(),
        label='Site (ID)',
    )
```
